src/core_plugins/timelion/public/directives/timelion_expression_input_helpers.js

```diff
@@ -52,13 +69,69 @@ export class FunctionSuggestions {
   }
 }
 
-export function suggest(expression, functionList, Parser) {
+function inLocation(cursorPosition, location) {
+  return cursorPosition >= location.min && cursorPosition <= location.max;
+}
+
+function extractSuggestionsFromParsed(result, cursorPosition, functionList) {
+  const activeFunc = result.functions.find((func) => {
+    return cursorPosition >= func.location.min && cursorPosition < func.location.max;
+  });
+
+  if (activeFunc) {
+    const funcDefinition = functionList.find((func) => {
+      return func.name === activeFunc.function;
+    });
+    const alreadyProvidedArguments = activeFunc.arguments.map((arg) => {
+      return arg.name;
+    });
+
+    // return function suggestion if cursor is outside of parentheses
+    // location range includes '.', function name, and '('.
+    const openParen = activeFunc.location.min + activeFunc.function.length + 2;
+    if (cursorPosition < openParen) {
+      return { list: [funcDefinition], location: activeFunc.location, type: SUGGESTION_TYPE.FUNCTIONS };
+    }
+
+    const args = funcDefinition.chainable ? funcDefinition.args.slice(1) : funcDefinition.args.slice(0);
+    const activeArg = activeFunc.arguments.find((argument) => {
+      return inLocation(cursorPosition, argument.location);
+    });
+
+    // return argument_value suggestions when cursor is inside agrument value
+    if (activeArg && activeArg.type === 'namedArg' && inLocation(cursorPosition, activeArg.value.location)) {
+      return { list: [], location: activeArg.value.location, type: SUGGESTION_TYPE.ARGUMENT_VALUE };
```

Comment:
It seems like this isn't actually returning anything valuable. Am I missing something? Why not return `undefined` here, and allow the caller to deal with it appropriately?

Reply:
you are right. In this PR it returns nothing valuable. But it lays the ground work for argument_value suggestions - when those get implemented in another PR